docs/design/2022-04-21-performance-diagnosis-enhancements.md

```diff
@@ -0,0 +1,188 @@
+- GC placement and label rules
+- Load data for cached tables
+- Alloc global ID for some operations
+    - Write sequence update to binlog
+    - Create temporary table
+
+When TiDB bootstrapping, it’ll create some internal sessions, these sessions are used for:
+
+- Bind info read/write
+- Privilege read/write
+- System variable read/write
+- Telemetry report
+- Statistic load
+
+The coprocessor request, transaction, and snapshot created by `RunInNewTxn` and the internal sessions need to be marked as internal.
+
+#### kvproto
+
+Add a field `request_source` in the proto, which is a string type field and can be empty(TiDB does not allow empty `request_source`). The string type can make the metrics display more readable.
+
+```diff
+// Miscellaneous metadata attached to most requests.
+message Context {
+    reserved 4;
+    ...
++   // The source of the request.
++   string request_source = xx;
+}
+
+```
+
+In TiDB, `request_source` is concatenated by **scope** and **type**, the scope is one of "internal" and "external", for type, there will be some variation between internal and external transactions.
+
+The type of the internal transactions would be their capabilities, the internal transactions are listed in the [Internal Transaction Usage](#internal-transaction-usage) section. The concatenated `request_source` would be like "internal_ddl_backfill_index", "internal_bindinfo_load", "internal_statistic_load", etc.
+
+The type of external transaction is the statement type. The concatenated `request_source` would be like "external_select", "external_insert", "external_commit", etc.
+
+#### Session
+
+Add a flag in `SessionVars`, TiDB creates some internal sessions when bootstrap, and the internal flags are set after bootstrap.
+
+```diff
+type SessionVars struct {
+	...
++	Internal bool
+}
+```
+
+#### Coprocessor
+
+```diff
+// Request represents a kv request.
+type Request struct {
+	...
++	RequestSourceScope string
++	RequestSourceType string
+}
+```
+
+#### KV Client
+
+Add in transaction and snapshot.
+
+```diff
+// KVTxn contains methods to interact with a TiKV transaction.
+type KVTxn struct {
+	...
++	RequestSourceScope string
++	RequestSourceType string
+}
+```
+
+```diff
+// KVSnapshot implements the tidbkv.Snapshot interface.
+type KVSnapshot struct {
+	...
++	RequestSourceScope string
++	RequestSourceType string
+}
+```
+
+The default value of `internal` is `false`(from the user client), and we should mark TiDB’s internal transaction as `true`.
+
+#### Metrics in TiDB
+
+After the internal and external transactions are differentiated, we can attach the flag to metrics of requests and KV client.
+
+Metrics need to be collected in both internal and external ways:
+
+- The metrics in read requests
+    - Get
+    - BatchGet
+    - Coprocessor
+    - Scan
+    - UnionScan
+- The 2PC related metrics
+    - Pessimistic Lock
+    - Prewrite
+    - Commit
+
+Metrics collected for external transactions only:
+
+- UnionScan
```

Comment:
Already appeared in above 'both internal and external' list, remove here ?

Reply:
Removed.